smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,457 @@
+
+        # selecting the 1st budget for 1st round of successive halving
+        sh_iter = self.max_sh_iter
+        budget = self.max_budget / (self.eta ** sh_iter)
+        budget = max(budget, self.min_budget)
+        prev_budget = 0
+
+        first_run = True
+
+        self.logger.debug('---' * 40)
+        self.logger.debug('Successive Halving run begins.')
+
+        # run intensification till budget is max
+        while budget <= self.max_budget:
+
+            self.logger.info('Running with budget [%.2f / %d] with %d challengers' % (budget, self.max_budget,
+                                                                                      len(curr_challengers)))
+            # selecting instance subset for this budget, depending on the kind of budget
+            available_insts = all_instances[int(prev_budget):int(budget)] if not self.cutoff_as_budget \
+                else all_instances
+
+            # determine 'k' for the next iteration
+            next_n_chal = int(np.round(len(curr_challengers) / self.eta))
+
+            try:
+                # Race all challengers
+                curr_challengers = self._race_challengers(challengers=curr_challengers,
+                                                          incumbent=incumbent,
+                                                          instances=available_insts,
+                                                          run_history=run_history,
+                                                          budget=budget,
+                                                          inc_sum_cost=inc_sum_cost,
+                                                          first_run=first_run)
+
+                # if all challengers were capped, then stop intensification
+                if not curr_challengers:
+                    self.logger.info("All configurations have been eliminated!"
+                                     "Interrupting optimization run and returning current incumbent")
+                    inc_perf = run_history.get_cost(incumbent)
+                    return incumbent, inc_perf
+
+            except BudgetExhaustedException:
+                # Returning the final incumbent selected so far because we ran out of optimization budget
+                self.logger.debug("Budget exhausted; "
+                                  "Interrupting optimization run and returning current incumbent")
+                inc_perf = run_history.get_cost(incumbent)
+                return incumbent, inc_perf
+
+            # selecting the top 'k' challengers for the next iteration
+            if next_n_chal > 0:
+                curr_challengers = self._top_k(curr_challengers, run_history, k=next_n_chal)
+            else:
+                # returning at least 1 configuration for next iteration since the incumbent has to run on all instances
+                self.logger.info('Ran out of configurations! Returning the best from what was seen so far.')
```

Comment:
I find this message counterintuitive as it so far was only emitted after successive halving successfully raced all configurations. I think this message should only be emitted in case that the number of challengers is unequal to the natural number of things to compare against each other